nmostesting/mocks/Registry.py

```diff
@@ -271,3 +404,184 @@ def heartbeat(version, node_id):
+            resource_id = request.args.get('id')
+            base_data.append(registry.get_resources()[resource_type][resource_id])
+        else:
+            data = registry.get_resources()[resource_type]
+
+            # only paginate for verison v1.3 and up 
+            if NMOSUtils.compare_api_version("v1.3", version) < 0:
+                for key, value in data.items():
+                    base_data.append(value)
+            else:
+                data_list = [d for k, d in data.items()]
+                sorted_list = sorted(data_list, key=functools.cmp_to_key(compare_resources))
+
+                result_count = 0
+                for value in sorted_list:
+                    if NMOSUtils.compare_resource_version(value['version'], since) >= 0 and NMOSUtils.compare_resource_version(until, value['version']) > 0:
+                        if result_count < registry.paging_limit:
+                            base_data.append(value)
+                            result_count += 1
+                        else:
+                            new_until = value['version']
+                            break
+
+    except Exception:
+        pass
+
+    response = Response(json.dumps(base_data), mimetype='application/json')
+
+    # add pagination headers for v1.3 and up
+    if NMOSUtils.compare_api_version("v1.3", version) >= 0:
+        link = ""
+        if new_until != MAX_UNTIL:
+            link += "<http://localhost:5102/x-nmos/query/v1.3/" + resource_type + "s/?paging.since=" + new_until + "&paging.limit=" + str(registry.paging_limit) + ">; rel=\"next\""
+        
+        if since != MIN_SINCE:
+            if link != "":
+                link += ","
+            link += "<http://localhost:5102/x-nmos/query/v1.3/" + resource_type + "s/?paging.since=0:0&paging.limit=" + str(registry.paging_limit) + ">; rel=\"first\""
+
+        if link != "": 
+            response.headers["Link"] = link
+        response.headers["X-Paging-Limit"] = registry.paging_limit 
+        response.headers["X-Paging-Since"] = since
+        response.headers["X-Paging-Until"] = new_until
+        response.headers["X-Ancestry-Generations"] = 1
+
+    return response
+
+
+@REGISTRY_API.route('/x-nmos/query/<version>/<resource>/<resource_id>', methods=['GET'], strict_slashes=False)
+def get_resource(version, resource, resource_id):
+    registry = REGISTRIES[flask.current_app.config["REGISTRY_INSTANCE"]]
+    if not registry.enabled:
+        abort(503)
+    authorized = registry.check_authorized(request.headers, request.path)
+    if authorized is not True:
+        abort(authorized)
+
+    registry.query_api_called = True
+
+    resource_type = resource.rstrip("s")
+    data = []
+    try:
+        # Type may not be in the list, so this could throw an exception
+        data = registry.get_resources()[resource_type][resource_id]
+    except Exception:
+        pass
+
+    return Response(json.dumps(data), mimetype='application/json')
```

Comment:
This should probably return a 404 is a client requests an ID which doesn't exist

Reply:
Yes - now corrected